k2/csrc/fsa_algo.h

```diff
@@ -91,71 +91,13 @@ bool Connect(const Fsa &a, Fsa *b, std::vector<int32_t> *arc_map = nullptr);
-     from those states, which will become the arcs in the output.  We'll also
-     store a back-pointer array that will allow us to figure out the best path
-     back to ki, in order to produce the output `arc_map`.    Assume we have
-     arrays
-
-     local_forward_weights (float) and local_backpointers (int) indexed by
-     state-id, and that the local_forward_weights are initialized with
-     -infinity's each time we process a new ki. (we have to figure out how to do
-     this efficiently).
-
-
-      Processing input-state ki:
-         local_forward_state_weights[ki] = forward_state_weights[ki] // from
-   WfsaWithFbWeights.
-                                                                     // Caution:
-   we should probably use
-                                                                     // double
-   here; these kinds of algorithms
-                                                                     // are
-   extremely sensitive to roundoff for
-                                                                     // very
-   long FSAs. local_backpointers[ki] = -1  // will terminate a sequence..
-         queue.push_back(ki)
-         while (!queue.empty()) {
-            ji = queue.front()  // we have to be a bit careful about order here,
-   to make sure
-                                // we always process states when they already
-   have the
-                                // best cost they are going to get.  If
-                                // FSA was top-sorted at the start, which we
-   assume, we could perhaps
-                                // process them in numerical order, e.g. using a
-   heap. queue.pop_front() for each arc leaving state ji: next_weight =
-   local_forward_state_weights[ji] + arc_weights[this_arc_index] if next_weight
-   + backward_state_weights[arc_dest_state] < best_path_weight - beam: if arc
-   label is epsilon: if next_weight < local_forward_state_weight[next_state]:
-                           local_forward_state_weight[next_state] = next_weight
-                           local_backpointers[next_state] = ji
-                else:
-                    add an arc to the output FSA, and create the appropriate
-                    arc_map entry by following backpointers (hopefully you can
-                    figure out the details).  Note: the output FSA's weights can
-   be computed later on, by calling code, using the info in arc_map.
+    @param [out] arc_derivs  Indexed by arc in `b`, this is the sequence of
+                      arcs in `a` that this arc in `b` corresponds to; the
+                      weight of the arc in b will equal the sum of those input
+                      arcs' weights
  */
 void RmEpsilonsPrunedMax(const WfsaWithFbWeights &a, float beam, Fsa *b,
-                         std::vector<std::vector<int32_t>> *arc_map);
+                         std::vector<std::vector<int32_t>> *arc_derivs);
```

Comment:
Can we keep `arc_map`?
I am thinking of arc derivatives once I see `arc_derivs`, though it has nothing
to do with derivatives.

Reply:
`arc_derivs` may be better, as  `arc_derivs` here actually is `vector<vector<pair<int32_t, int32_t>` where `pair.second` is always `1`. Another reason is that `map` usually means `1 to 1 `mapping in convention, especially in the perspective of function/algorithm, but `arc_derivs` here usually is a `one to many` mapping. Dan aslo uses this name in many other algorithms, such as determinize.

Reply:
mm, I prefer arc_derivs too.  it does actually represent derivatives in a sense, as in it says how the weights on the output arcs vary with those of the input arcs.